pkg/compact/compact.go

```diff
@@ -326,14 +330,34 @@ func (s *Syncer) garbageCollect(ctx context.Context, resolution int64) error {
 			return retry(errors.Wrapf(err, "delete block %s from bucket", id))
 		}
 
-		// Immediately update our in-memory state so no further call to SyncMetas is needed
-		// after running garbage collection.
-		delete(s.blocks, id)
+		// Mark the blocks as to be removed from the blocks map so that
+		// it is not considered for the next round of compaction.
+		s.blocksToDelete[id] = time.Now().UTC()
 		s.metrics.garbageCollectedBlocks.Inc()
 	}
 	return nil
 }
 
+func (s *Syncer) DeleteDelayFilter(metas map[ulid.ULID]*metadata.Meta, synced block.GaugeLabeled, _ bool) {
+	for id := range metas {
+		blockMarkedTime, ok := s.blocksToDelete[id]
+
+		if !ok {
+			continue
+		}
+
+		// Wait for consistencyDelay for the compacted block to be successfully uploaded to bucket.
+		if ulid.Now()-ulid.Timestamp(blockMarkedTime) < uint64(s.consistencyDelay) {
+			level.Debug(s.logger).Log("msg", "block has been marked for deletion", "block", id)
+			synced.WithLabelValues(block.MarkedForDeletion).Inc()
+			delete(metas, id)
+		} else {
+			// If the compacted block is successfully uploaded, remove the old block from the group.
+			delete(s.blocks, id)
```

Comment:
So we remove this from group.. but don't we want actually want to remove the block from the bucket? Also doing here is weird. It is a filter of metaFetcher. Filter can be run any time. I believe meta Fetcher filter might be capable to NOT return blocks that are scheduled to be removed, but BASED on Source ULID on meta.json not on some map stored in memory. This is because when compactor crashes this information will be lost.

Let's talk about this offline we can setup a call (: